src/rust/inetstack/protocols/tcp/passive_open.rs

```diff
@@ -140,44 +148,33 @@ impl<const N: usize> SharedPassiveSocket<N> {
         self.ready.pop(&yielder).await?
     }
 
-    /// Receive and direct new connection requests and ACKs.
-    pub fn receive(&mut self, ip_header: &Ipv4Header, header: TcpHeader, buf: DemiBuffer) -> Result<(), Fail> {
-        let remote = SocketAddrV4::new(ip_header.get_src_addr(), header.src_port);
-        // If the packet is for an inflight connection, route it there.
-        if let Some(inflight) = self.inflight.remove(&remote) {
-            // If ack for inflight connection request, remove from inflight table and handle. If error, we do not put
-            // the inflight request back but drop it.
-            // FIXME: https://github.com/microsoft/demikernel/issues/1054
-            if header.ack {
-                return self.handle_ack(inflight, remote, header, buf);
-            } else {
-                return Err(Fail::new(EBADMSG, "expecting ACK"));
-            }
-        } else {
-            // Check whether this packet is for a connection that we have finished accepting.
-            for result in self.ready.get_mut_values() {
-                match result {
-                    // We've finished establishing the connection, so just deliver the packet.
-                    Ok(ref mut socket) if socket.endpoints().1 == remote => {
-                        socket.receive(header, buf);
-                        return Ok(());
-                    },
-                    _ => continue,
-                }
+    async fn poll(mut self, yielder: Yielder) {
+        loop {
+            let (ipv4_hdr, tcp_hdr, buf) = match self.recv_queue.pop(&yielder).await {
+                Ok(result) => result,
+                Err(_) => break,
+            };
+            let remote = SocketAddrV4::new(ipv4_hdr.get_src_addr(), tcp_hdr.src_port);
+            if let Some(recv_queue) = self.connections.get_mut(&remote) {
+                // Packet is either for an inflight request or established connection.
+                recv_queue.push((ipv4_hdr, tcp_hdr, buf));
+                continue;
             }
-        }
 
-        // If not a SYN, then this packet is not for a new connection and we throw it away.
-        if !header.syn || header.ack || header.rst {
-            return Err(Fail::new(EBADMSG, "invalid flags"));
+            // If not a SYN, then this packet is not for a new connection and we throw it away.
+            if !tcp_hdr.syn || tcp_hdr.ack || tcp_hdr.rst {
+                let cause = "invalid flags";
+                warn!("{}", cause);
+                continue;
+            }
```

Comment:
We should send a reset on this case, I will change https://github.com/microsoft/demikernel/pull/1074 for doing so.

Reply:
Perfect, thanks!